src/odbc_connection.cpp

```diff
@@ -2123,44 +2124,72 @@ SQLRETURN ODBCConnection::FetchAll(QueryData *data) {
 
     // Iterate over each column, putting the data in the row object
     for (int i = 0; i < data->columnCount; i++) {
-      row[i].size = data->columns[i]->StrLen_or_IndPtr;
-      if (row[i].size == SQL_NULL_DATA || row[i].size == SQL_NO_TOTAL) {
-        row[i].char_data = NULL;
-      } else {
-        switch (data->columns[i]->bind_type) {
-          case SQL_C_DOUBLE:
-            row[i].double_data = *(SQLDOUBLE*)data->boundRow[i];
-            break;
 
-          case SQL_C_UTINYINT:
-            row[i].tinyint_data = *(SQLCHAR*)data->boundRow[i];
-            break;
+      if (data->columns[i]->StrLen_or_IndPtr == SQL_NULL_DATA) {
+        row[i].size = SQL_NULL_DATA;
+      }
 
-          case SQL_C_USHORT:
-            row[i].smallint_data = *(SQLUSMALLINT*)data->boundRow[i];
-            break;
+      switch (data->columns[i]->bind_type) {
+        case SQL_C_DOUBLE:
+          row[i].double_data = *(SQLDOUBLE*)data->boundRow[i];
+          break;
 
-          case SQL_C_SLONG:
-            row[i].integer_data = *(SQLINTEGER*)data->boundRow[i];
-            break;
+        case SQL_C_UTINYINT:
+          row[i].tinyint_data = *(SQLCHAR*)data->boundRow[i];
+          break;
 
-          case SQL_C_UBIGINT:
-            row[i].ubigint_data = *(SQLUBIGINT*)data->boundRow[i];
-            break;
+        case SQL_C_USHORT:
+          row[i].smallint_data = *(SQLUSMALLINT*)data->boundRow[i];
+          break;
 
-          case SQL_C_WCHAR:
-            row[i].wchar_data = new SQLWCHAR[row[i].size + 2]();
-            memcpy(row[i].wchar_data, data->boundRow[i], row[i].size);
-            break;
+        case SQL_C_SLONG:
+          row[i].integer_data = *(SQLINTEGER*)data->boundRow[i];
+          break;
 
-          case SQL_C_CHAR:
-          default:
-            row[i].char_data = new SQLCHAR[row[i].size + 2]();
-            memcpy(row[i].char_data, data->boundRow[i], row[i].size);
-            break;
-        }
-        row->bind_type = data->columns[i]->bind_type;
+        case SQL_C_UBIGINT:
+          row[i].ubigint_data = *(SQLUBIGINT*)data->boundRow[i];
+          break;
+
+        case SQL_C_BINARY:
+          row[i].char_data = new SQLCHAR[row[i].size]();
+          memcpy(row[i].char_data, data->boundRow[i], row[i].size);
+          break;
+
+        case SQL_C_WCHAR:
+          row[i].size = wcslen((const wchar_t *)data->boundRow[i]) * sizeof(SQLWCHAR);
+          row[i].wchar_data = new SQLWCHAR[(row[i].size + 2) * sizeof(SQLWCHAR)]();
+          memcpy(row[i].wchar_data, data->boundRow[i], row[i].size);
+          break;
+
+        case SQL_C_CHAR:
+        default:
+          row[i].size = strlen((const char *)data->boundRow[i]);
+          row[i].char_data = new SQLCHAR[(row[i].size + 2) * sizeof(SQLCHAR)]();
+          memcpy(row[i].char_data, data->boundRow[i], row[i].size);
+          break;
```

Comment:
If the data was returned as NULL, row[i].size will be SQL_NULL_DATA, so these will crash.

The double/short/bigint/whatever code above will also give undefined results when the data is SQL_NULL_DATA, since the data in the pointer is undefined. Passing random data through the floating point unit when reinterpreting as double (ie. `*(double*)`) can also cause floating point exceptions.

Reply:
Yeah the switch statement should be in an "else" with the `SQL_NULL_DATA`. Then when the data is converted to Napi values, the first thing it does is check this size to see if it is `SQL_NULL_DATA` and then assign it a JavaScript null value:

https://github.com/markdirish/node-odbc/blob/fca28179f231782d6db008f1433afa027e64dd5c/src/odbc_connection.cpp#L330-L335